server/shared/src/main/scala/org/http4s/server/middleware/LoggerConfig.scala

```diff
@@ -0,0 +1,154 @@
+      logAction: F[String => F[Unit]],
+  ): LoggerConfig[F] =
+    new LoggerConfig[F](logHeaders, logBody, redactHeadersWhen, logAction) {}
+
+  /** Creates a builder for LoggerConfig.
+    */
+  def default[F[_]: Async]: LoggerConfigBuilder[F] =
+    LoggerConfigBuilder[F](
+      logHeaders = false,
+      logBody = false,
+      redactHeadersWhen = Logger.defaultRedactHeadersWhen,
+      logAction = Async[F].pure(s => logger.info(s).to[F]),
+    )
+
+  object LoggerConfigBuilder {
+    private[LoggerConfig] def apply[F[_]: Async](
+        logHeaders: Boolean,
+        logBody: Boolean,
+        redactHeadersWhen: CIString => Boolean,
+        logAction: F[String => F[Unit]],
+    ): LoggerConfigBuilder[F] =
+      new LoggerConfigBuilder[F](logHeaders, logBody, redactHeadersWhen, logAction) {}
+  }
+
+  sealed abstract case class LoggerConfigBuilder[F[_]: Async] private[LoggerConfig] (
+      logHeaders: Boolean,
+      logBody: Boolean,
+      redactHeadersWhen: CIString => Boolean,
+      logAction: F[String => F[Unit]],
+  ) {
+    private def copy(
+        logHeaders: Boolean = this.logHeaders,
+        logBody: Boolean = this.logBody,
+        redactHeadersWhen: CIString => Boolean = this.redactHeadersWhen,
+        logAction: F[String => F[Unit]] = this.logAction,
+    ): LoggerConfigBuilder[F] =
+      LoggerConfigBuilder(logHeaders, logBody, redactHeadersWhen, logAction)
+
+    /** Sets whether to log headers.
+      *
+      * @param logHeaders
+      */
+    def withLogHeaders(logHeaders: Boolean): LoggerConfigBuilder[F] =
+      copy(logHeaders = logHeaders)
+
+    /** Sets whether to log body.
+      * @param logBody
+      */
+    def withLogBody(logBody: Boolean): LoggerConfigBuilder[F] =
+      copy(logBody = logBody)
+
+    /** Sets the function to determine which headers should be redacted.
+      *
+      * @param redactHeadersWhen function to determine which headers to redact
+      */
+    def withRedactHeadersWhen(redactHeadersWhen: CIString => Boolean): LoggerConfigBuilder[F] =
+      copy(redactHeadersWhen = redactHeadersWhen)
+
+    /** Sets the log action function.
+      *
+      * @param f log action
+      */
+    def withLogAction(action: String => F[Unit]): LoggerConfigBuilder[F] =
+      copy(logAction = Async[F].pure(action))
+
+    /** Sets a deferred log action function.
+      *
+      * Use this when you want loggerFactory to capture context e.g. via IOLocal
+      * at the time of logging. Otherwise, context may be lost.
+      * Important for cases where `logBody` is enabled, which may happen on
+      * another fiber.
+      *
+      * @example  {{{
+      * import cats.effect.IOLocal
+      * import org.typelevel.log4cats.Logger
+      * val ioLocal = IOLocal(Map.empty[String, String])
+      * val logger: Logger[IO] = ...
+      * val logF: IO[String => IO[Unit]] =
+      *   ioLocal.get.map { ctx =>
+      *     { msg => logger.info("log msg and ctx: ...") }
```

Comment:
String interpolating the values for the example fails doc generation on native and JS